src/wmtk/simplex/link_condition.cpp

```diff
@@ -98,4 +98,93 @@ bool link_condition(const TriMesh& mesh, const Tuple& edge)
+    const Simplex v_b = Simplex::vertex(mesh.switch_tuple(edge, PrimitiveType::Vertex));
+    const Simplex e_ab = Simplex::edge(edge);
+    const SimplexCollection link_a = link(mesh, v_a); // link(a)
+    const SimplexCollection link_b = link(mesh, v_b); // link(b)
+    const SimplexCollection link_ab = link(mesh, e_ab); // link(ab)
+
+    const SimplexCollection link_a_link_b_intersection =
+        SimplexCollection::get_intersection(link_a, link_b);
+
+    return SimplexCollection::are_simplex_collections_equal(link_a_link_b_intersection, link_ab);
+}
+
+bool link_condition(const TetMesh& mesh, const Tuple& edge)
+{
+    // close mesh check
+    if (!link_condition_closed_tetmesh(mesh, edge)) {
+        return false;
+    }
+
+    // add dummy vertex and link it to boundary faces
+    // the check can be convert from link_dummy(a) \intersect link_dummy(b) == link_dummy(ab) to:
+    // if ab is a boundary edge, then if any boundary face incident a shares a common edge with any
+    // boundary face incident b, return false.
+    // if ab is not boundary edge, then if a is not on
+    // boundary or b is not on boundary, return true, otherwise false
+
+    auto get_boundary_faces = [&mesh](const Tuple& _v) {
+        const Simplex input_v(PrimitiveType::Vertex, _v);
+        std::vector<Tuple> ret;
+        // get incident_faces from open_star
+        auto incident_faces = open_star(mesh, input_v).simplex_vector(PrimitiveType::Face);
+        for (const Simplex& _f : incident_faces) {
+            if (mesh.is_boundary(_f.tuple(), PrimitiveType::Face)) {
+                if (utils::SimplexComparisons::equal(
+                        mesh,
+                        Simplex(PrimitiveType::Vertex, _f.tuple()),
+                        input_v)) {
+                    // tuple edge is _v - x
+                    ret.push_back(mesh.switch_edge(mesh.switch_vertex(_f.tuple())));
+                } else {
+                    if ((utils::SimplexComparisons::equal(
+                            mesh,
+                            Simplex(PrimitiveType::Vertex, mesh.switch_vertex(_f.tuple())),
+                            input_v))) {
+                        // tuple edge is x - _v
+                        ret.push_back(mesh.switch_edge(_f.tuple()));
+                    } else {
+                        // tuple edge is x - y
+                        ret.push_back(_f.tuple());
+                    }
+                }
+            }
+        }
+        return ret;
+    };
+
+    const auto boundary_neighbors_a = get_boundary_faces(edge);
+    const auto boundary_neighbors_b =
+        get_boundary_faces(mesh.switch_tuple(edge, PrimitiveType::Vertex));
+    if (mesh.is_boundary_edge(edge)) {
+        for (auto f_a : boundary_neighbors_a) {
+            for (auto f_b : boundary_neighbors_b) {
+                if (utils::SimplexComparisons::equal(
+                        mesh,
+                        Simplex(PrimitiveType::Edge, f_a),
```

Comment:
I am very confused. In `get_boundary_faces` you are collecting triangles, right? Then why do you now consider the triangle here as an edge?
I think you should use `Simplex` more often to make the code easier to read. I cannot really follow the reasoning of that code right now.

Reply:
The reason is I want to check: after adding a dummy vertex and connecting it to all of the vertices in the incident boundary faces of v_a and v_b, do link(a) and link(b) share new common faces (one of the endpoint should be the dummy vertex). So the real check is to check if any face from boundary_neighbors_a shares a common edge with any face from boundary_neighbors_b, because if they shares an edge, there will be a common face consisting of this edge and the dummy vertex sharing between link(a) and link(b), which does not belongs to link(ab).